// igniteui-react-charts/esm2015/lib/AssigningCategoryStyleEventArgs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AssigningCategoryStyleEventArgsBase } from "./AssigningCategoryStyleEventArgsBase";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let AssigningCategoryStyleEventArgs = /*@__PURE__*/ (() => {
    class AssigningCategoryStyleEventArgs extends AssigningCategoryStyleEventArgsBase {
        constructor() {
            super(...arguments);
            this.r = 0;
            this.s = null;
            this.t = 0;
            this.p = 0;
            this.q = 0;
        }
    }
    AssigningCategoryStyleEventArgs.$t = /*@__PURE__*/ markType(AssigningCategoryStyleEventArgs, 'AssigningCategoryStyleEventArgs', AssigningCategoryStyleEventArgsBase.$);
    return AssigningCategoryStyleEventArgs;
})();
