// igniteui-react-charts/esm2015/lib/CrosshairLayerFrame.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AxisAnnotationFrame } from "./AxisAnnotationFrame";
import { List$1 } from "igniteui-react-core";
import { Brush } from "igniteui-react-core";
import { Frame } from "./Frame";
import { FrameExtended } from "./FrameExtended";
import { Number_$type, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let CrosshairLayerFrame = /*@__PURE__*/ (() => {
    class CrosshairLayerFrame extends AxisAnnotationFrame {
        constructor() {
            super();
            this.ad = null;
            this.ae = null;
            this.ac = null;
            this.af = null;
            this.ad = new List$1(Number_$type, 0);
            this.ae = new List$1(Number_$type, 0);
            this.ac = new List$1(Brush.$, 0);
            this.af = new List$1(Brush.$, 0);
        }
        c(a, b, c) {
            super.c(a, b, c);
            let d = b;
            let e = c;
            Frame.b(this.ad, a, d.ad, e.ad);
            Frame.b(this.ae, a, d.ae, e.ae);
            FrameExtended.a(this.ac, a, d.ac, e.ac, 0);
            FrameExtended.a(this.af, a, d.af, e.af, 0);
        }
    }
    CrosshairLayerFrame.$t = /*@__PURE__*/ markType(CrosshairLayerFrame, 'CrosshairLayerFrame', AxisAnnotationFrame.$);
    return CrosshairLayerFrame;
})();
