// igniteui-react-charts/esm2015/lib/DictInterpolator$3.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, fromEnum, markType } from "igniteui-react-core";
import { List$1 } from "igniteui-react-core";
/**
 * @hidden
 */
export let DictInterpolator$3 = /*@__PURE__*/ (() => {
    class DictInterpolator$3 extends Base {
        constructor($tKey, $tValue, $tFrame, a, b, c, d) {
            super();
            this.$tKey = null;
            this.$tValue = null;
            this.$tFrame = null;
            this.a = null;
            this.c = null;
            this.d = null;
            this.b = null;
            this.$tKey = $tKey;
            this.$tValue = $tValue;
            this.$tFrame = $tFrame;
            this.$type = this.$type.specialize(this.$tKey, this.$tValue, this.$tFrame);
            this.a = a;
            this.c = b;
            this.d = c;
            this.b = d;
        }
        e(a, b, c, d, e, f) {
            let g = new List$1(this.$tKey, 0);
            for (let h of fromEnum(a.keys)) {
                let i;
                let j;
                let k = a.item(h);
                let l = ((() => { let m = c.tryGetValue(h, i); i = m.p1; return m.ret; })());
                let m = ((() => { let n = d.tryGetValue(h, j); j = n.p1; return n.ret; })());
                if (!l && !m) {
                    g.add(h);
                }
                else {
                    this.a(k, b, i, j, e, f);
                    if (!this.d(k)) {
                        g.add(h);
                    }
                }
            }
            for (let n of fromEnum(g)) {
                a.removeItem(n);
            }
            for (let o of fromEnum(c.keys)) {
                let p = c.item(o);
                let q;
                let r;
                let s = d.tryGetValue(o, q);
                q = s.p1;
                let t = ((() => { let u = a.tryGetValue(o, r); r = u.p1; return u.ret; })());
                if (!t) {
                    r = this.b();
                    this.a(r, b, p, q, e, f);
                    if (!this.d(r)) {
                        continue;
                    }
                    a.addItem(this.c(r), r);
                }
            }
            for (let u of fromEnum(d.keys)) {
                let v = d.item(u);
                let w;
                let x;
                let y = ((() => { let z = c.tryGetValue(u, w); w = z.p1; return z.ret; })());
                let z = ((() => { let aa = a.tryGetValue(u, x); x = aa.p1; return aa.ret; })());
                if (!z && !y) {
                    x = this.b();
                    this.a(x, b, w, v, e, f);
                    if (!this.d(x)) {
                        continue;
                    }
                    a.addItem(this.c(x), x);
                }
            }
        }
    }
    DictInterpolator$3.$t = /*@__PURE__*/ markType(DictInterpolator$3, 'DictInterpolator$3');
    return DictInterpolator$3;
})();
