// igniteui-react-charts/esm2015/lib/igr-callout-label-updating-event-args.js
import { CalloutLabelUpdatingEventArgs as CalloutLabelUpdatingEventArgs_internal } from "./CalloutLabelUpdatingEventArgs";
/**
 * Represents event arguments for updating label of callout layer
*/
export class IgrCalloutLabelUpdatingEventArgs {
    createImplementation() {
        return new CalloutLabelUpdatingEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets data X-value assositated with callout layer
    */
    get xValue() {
        return this.i.c;
    }
    set xValue(v) {
        this.i.c = v;
    }
    /**
     * Gets data Y-value assositated with callout layer
    */
    get yValue() {
        return this.i.d;
    }
    set yValue(v) {
        this.i.d = v;
    }
    /**
     * Gets data item assositated with callout layer
    */
    get item() {
        return this.i.item;
    }
    set item(v) {
        this.i.item = v;
    }
    /**
     * Gets or sets label displayed by callout layer
    */
    get label() {
        return this.i.b;
    }
    set label(v) {
        this.i.b = v;
    }
}
