// igniteui-react-charts/esm2015/lib/igr-slice-click-event-args.js
import { ensureBool } from "igniteui-react-core";
/**
 * Contains PieChart click event data.
*/
export class IgrSliceClickEventArgs {
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * Gets or sets whether the slice is selected.
    */
    get isSelected() {
        return this.i.isSelected;
    }
    set isSelected(v) {
        this.i.isSelected = ensureBool(v);
    }
    /**
     * Gets or sets whether the slice is exploded.
    */
    get isExploded() {
        return this.i.isExploded;
    }
    set isExploded(v) {
        this.i.isExploded = ensureBool(v);
    }
    /**
     * Gets whether the current slice is part of the others slice.
    */
    get isOthersSlice() {
        return this.i.isOthersSlice;
    }
    /**
     * Gets the index of the slice that was clicked.
    */
    get index() {
        return this.i.index;
    }
    /**
     * Gets the slice data context.
    */
    get dataContext() {
        return this.i.dataContext;
    }
    /**
     * Holds a reference to the original event that triggered slice click.
    */
    get originalEvent() {
        return this.i.originalEvent;
    }
    set originalEvent(v) {
        this.i.originalEvent = v;
    }
}
