// igniteui-react-charts/esm2015/lib/igr-numeric-angle-axis.js
import { IgrNumericAxisBase } from "./igr-numeric-axis-base";
import { NumericAngleAxis } from "./NumericAngleAxis";
/**
 * Represents a IgxDataChartComponent angle based axis for polar series.
*/
export class IgrNumericAngleAxis extends IgrNumericAxisBase {
    createImplementation() {
        return new NumericAngleAxis();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets if the current axis is of angular axis type
    */
    get isAngular() {
        return this.i.b2;
    }
    /**
     * Indicates the angle in degress that the chart's 0th angle should be offset.
    */
    get startAngleOffset() {
        return this.i.lt;
    }
    set startAngleOffset(v) {
        this.i.lt = +v;
    }
    /**
     * Gets the scaled angle value in radians based on the raw input.
    
    * @param unscaledValue  * The raw input value.
    */
    getScaledAngle(unscaledValue) {
        let iv = this.i.getScaledAngle(unscaledValue);
        return (iv);
    }
    /**
     * Gets the raw axis value back from the angle that would be used on the chart.
    
    * @param scaledValue  * The chart angle value.
    */
    getUnscaledAngle(scaledValue) {
        let iv = this.i.getUnscaledAngle(scaledValue);
        return (iv);
    }
}
