// igniteui-react-charts/esm2015/lib/RingSeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RingSeriesBase } from "./RingSeriesBase";
import { Ring } from "./Ring";
import { ArcItem } from "./ArcItem";
import { fromEnum, markType } from "igniteui-react-core";
import { RingCollection } from "./RingCollection";
/**
 * @hidden
 */
export let RingSeries = /*@__PURE__*/ (() => {
    class RingSeries extends RingSeriesBase {
        constructor() {
            super();
            this.el = null;
            this.el = ((() => {
                let $ret = new Ring();
                $ret.c = this;
                $ret.e = false;
                return $ret;
            })());
            let a = ((() => {
                let $ret = new ArcItem();
                $ret.k = this.a8;
                $ret.d = this.el;
                $ret.n = this.cp;
                $ret.c = this.ap;
                $ret.j = this.a6;
                return $ret;
            })());
            this.el.g.add(a);
            this.ab = RingSeries.$;
        }
        c1(a, b, c, d) {
            super.c1(a, b, c, d);
            if (b == RingSeriesBase.$$p[6]) {
                for (let e = 0; e < this.el.ringControl.arcs.count; e++) {
                    this.el.ringControl.arcs.item(e).dq = d;
                }
            }
            if (b == RingSeriesBase.$$p[7]) {
                for (let f = 0; f < this.el.ringControl.arcs.count; f++) {
                    this.el.ringControl.arcs.item(f).ag = d;
                }
            }
            if (b == RingSeriesBase.$$p[28]) {
                for (let g = 0; g < this.el.ringControl.arcs.count; g++) {
                    this.el.ringControl.arcs.item(g).ek = d;
                }
            }
            if (b == RingSeriesBase.$$p[29]) {
                for (let h = 0; h < this.el.ringControl.arcs.count; h++) {
                    this.el.ringControl.arcs.item(h).aj = d;
                }
            }
            if (b == RingSeriesBase.$$p[18]) {
                for (let i = 0; i < this.el.ringControl.arcs.count; i++) {
                    this.el.ringControl.arcs.item(i).d6 = d;
                }
            }
            if (b == RingSeriesBase.$$p[19]) {
                for (let j = 0; j < this.el.ringControl.arcs.count; j++) {
                    this.el.ringControl.arcs.item(j).ah = d;
                }
            }
            if (b == RingSeriesBase.$$p[21]) {
                for (let k = 0; k < this.el.ringControl.arcs.count; k++) {
                    this.el.ringControl.arcs.item(k).eb = d;
                }
            }
            if (b == RingSeriesBase.$$p[22]) {
                for (let l = 0; l < this.el.ringControl.arcs.count; l++) {
                    this.el.ringControl.arcs.item(l).ai = d;
                }
            }
            if (b == RingSeriesBase.$$p[2]) {
                for (let m = 0; m < this.el.ringControl.arcs.count; m++) {
                    this.el.ringControl.arcs.item(m).ax = d;
                }
            }
            if (b == RingSeriesBase.$$p[3]) {
                for (let n = 0; n < this.el.ringControl.arcs.count; n++) {
                    this.el.ringControl.arcs.item(n).ay = d;
                }
            }
            if (b == RingSeriesBase.$$p[20]) {
                for (let o = 0; o < this.el.ringControl.arcs.count; o++) {
                    this.el.ringControl.arcs.item(o).d9 = this.b1;
                }
            }
            if (b == RingSeriesBase.$$p[0]) {
                this.cz();
            }
            if (b == RingSeriesBase.$$p[33]) {
                if (this.el.g != null && this.el.g.count > 0) {
                    this.el.g._inner[0].k = this.a8;
                }
            }
            if (b == RingSeriesBase.$$p[36]) {
                if (this.el.g != null && this.el.g.count > 0) {
                    this.el.g._inner[0].n = this.cp;
                    this.c0();
                }
            }
            if (this.el != null && this.aw.e()) {
                this.el.o();
                if (this.el.f()) {
                    this.el.c.aw.l();
                }
            }
        }
        au() {
            let a = new RingCollection();
            if (this.el.g._inner[0].g.count > 0) {
                a.add(this.el);
            }
            return a;
        }
        c0() {
            if (this.el != null) {
                this.el.g._inner[0].h = this.itemsSource;
                this.el.g._inner[0].p(this.a8);
                if (this.ax != null) {
                    this.ax.bx();
                    this.ax.b2();
                }
            }
        }
        cz() {
            if (this.el != null) {
                this.el.g._inner[0].a = this.ak;
            }
        }
        c5() {
            if (this.el != null) {
                this.width = this.el.q.width;
                this.height = this.el.q.height;
                this.aw.m(this.el.p.x, this.el.p.y);
            }
        }
        c4() {
            if (this.el != null) {
                for (let a of fromEnum(this.el.ringControl.arcs.a)) {
                    a.f0();
                }
            }
        }
        cv(a, b) {
            super.cv(a, b);
            if (this.el != null) {
                this.el.l(this.itemsSource, a, b);
            }
        }
        cu() {
            super.cu();
            if (this.el != null) {
                this.el.k(this.itemsSource);
            }
        }
        cw(a, b) {
            super.cw(a, b);
            if (this.el != null) {
                this.el.m(this.itemsSource, a, b);
            }
        }
        cx(a, b, c) {
            super.cx(a, b, c);
            if (this.el != null) {
                this.el.n(this.itemsSource, a, b, c);
            }
        }
    }
    RingSeries.$t = /*@__PURE__*/ markType(RingSeries, 'RingSeries', RingSeriesBase.$);
    return RingSeries;
})();
