// igniteui-react-charts/esm2015/lib/TimeAxisLabelFormat.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, INotifyPropertyChanged_$type, PropertyChangedEventArgs, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let TimeAxisLabelFormat = /*@__PURE__*/ (() => {
    class TimeAxisLabelFormat extends Base {
        constructor() {
            super(...arguments);
            this.e = (0);
            this.a = null;
            this.propertyChanged = null;
        }
        get f() {
            return this.e;
        }
        set f(a) {
            let b = this.f;
            if (a != b) {
                this.e = a;
                this.g("Range", b, this.f);
            }
        }
        get b() {
            return this.a;
        }
        set b(a) {
            let b = this.b;
            if (a != b) {
                this.a = a;
                this.g("Format", b, this.b);
            }
        }
        g(a, b, c) {
            this.h(a, b, c);
            if (this.propertyChanged != null) {
                this.propertyChanged(this, new PropertyChangedEventArgs(a));
            }
        }
        h(a, b, c) {
        }
    }
    TimeAxisLabelFormat.$t = /*@__PURE__*/ markType(TimeAxisLabelFormat, 'TimeAxisLabelFormat', Base.$, [INotifyPropertyChanged_$type]);
    return TimeAxisLabelFormat;
})();
